presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -160,11 +160,86 @@ protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void cont
             if (node.getValue() instanceof IsNullPredicate) {
                 return process(new IsNotNullPredicate(((IsNullPredicate) node.getValue()).getValue()));
             }
-            PlanNodeStatsEstimate sourceStats = process(node.getValue());
-            if (sourceStats.isOutputRowCountUnknown()) {
+            PlanNodeStatsEstimate source = process(node.getValue());
+            if (source.isOutputRowCountUnknown()) {
                 return PlanNodeStatsEstimate.unknown();
             }
-            return differenceInStats(input, sourceStats);
```

Comment:
Could you try abstracting new `subtract` method instead of inlining specific code?
It should be possible to create universal stats `subtract` method.

The problem we are solving here is multisets subtract. Each set has number of NDVs and total number of rows.
We know how many values there is per distinct value. Having this knowledge we can deduce how many NDVs are left after subtract operation (left - right).
Possibly such subtract method should be parametrized by `% of left NDVs covered by right NDVs`. Such factor can be passed directly or derived from ranges.

Reply:
I was thinking about that. But than i realized that it is used only in 2 places. When estimating `NOT` and `OR`, and the semantics are slightly different. For estimating `NOT` you need to only subtract. For estimating `OR` you need to `SUM` first and than subtract. Instead of doing 2 operations in estimating `OR` i think it is more readable to just inline both.

Reply:
Still it would be nice to put logical code blocks in separate methods. The code is easier to follow. People might change code in one place but forget about the other.

Reply:
If you don't feel strong about it, i would like to keep it inline for the sake of readability.

Reply:
> If you don't feel strong about it, i would like to keep it inline for the sake of readability.

Sure, we can inline it. The idea of `PlanNodeStatsEstimateMath` isn't bad though. It just needs it's semantics corrected so that those are true multiset operations (with proper parametrization)